proposals/0152-precompiles.md

```diff
@@ -0,0 +1,273 @@
+---
+simd: '0152'
+title: Precompiles
+authors:
+  - Emanuele Cesena
+category: Standard
+type: Core
+status: Draft
+created: 2024-06-03
+feature:
+supersedes:
+superseded-by:
+extends:
+---
+
+## Summary
+
+Define a unified behavior for precompiles, and highlight a few minor
+changes to the existing precompiles to minimize differences.
+
+## Motivation
+
+Precompiles are special native programs designed to verify additional
+transaction signatures.
+They run without the VM and without loading any account, and they can access
+data from other instructions, within the same transaction.
+
+At the time of writing, two precompiles exist to verify Ed25519 and
+Ethereum-like Secp256k1 signatures, and another one is being proposed to
+support Secp256r1 signatures for FIDO Passkeys.
+
+Historically, the two precompiles were built at different times and
+by different people, so naturally there are some subtle differences
+in how they behave, especially in edge cases.
+
+The main goal of this document is to provide a specification for how
+a precompile should behave, remove differences and provide guidelines
+for future proposals.
+
+In addition, we highlight 3 minor changes to the existing precompiles
+that will simplify their behavior and make it easier to develop
+alternative validator clients.
+
+## Alternatives Considered
+
+Leave as is.
+
+## New Terminology
+
+n/a
+
+## Detailed Design
+
+We first propose a specification for precompiles. Then we propose changes
+to the existing precompiles with respect to the specification.
+
+### Precompile Specification
+
+Precompiles are special native programs designed to verify additional
+signatures. Each precompile consists of a single `verify` instruction.
+
+Precompiles are executed right after transaction signature verification,
+they run without the VM and without loading any account.
+
+If a transaction contains more than 8 precompile signatures, it must fail.
+
+```
+MAX_ALLOWED_PRECOMPILE_SIGNATURES = 8
+```
+
+The precompile instruction `verify` accepts the following data:
+
+```
+struct PrecompileVerifyInstruction {
+  num_signatures:  u8,                       // Number of signatures to verify
+  padding:         u8,                       // Single byte padding
+  offsets:         Array<PrecompileOffsets>, // Array of offsets
+  additionalData?: Bytes,                    // Optional additional data, e.g.
+                                             // signatures included in the same
+                                             // instruction
+}
+
+struct PrecompileOffsets {
+  signature_offset: u16 LE,                  // Offset to signature
```

Comment:
Offset could be a bit more clear. It's the offset within specified instruction data, right?